actors/miner/src/vesting_state.rs

```diff
@@ -1,54 +1,120 @@
 // Copyright 2019-2022 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-use std::{iter, mem};
+use std::iter;
 
+use cid::Cid;
+use fil_actors_runtime::{ActorError, AsActorError};
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::serde::{Deserialize, Serialize};
 use fvm_ipld_encoding::tuple::*;
+use fvm_ipld_encoding::CborStore;
 use fvm_shared::clock::ChainEpoch;
 use fvm_shared::econ::TokenAmount;
-use itertools::{EitherOrBoth, Itertools};
+use fvm_shared::error::ExitCode;
+use itertools::{EitherOrBoth, Itertools, PeekingNext};
+use multihash_codetable::Code;
 use num_traits::Zero;
 
 use super::{QuantSpec, VestSpec};
 
 // Represents miner funds that will vest at the given epoch.
-#[derive(Debug, Serialize_tuple, Deserialize_tuple)]
+#[derive(Default, Debug, Serialize_tuple, Deserialize_tuple, Clone)]
 pub struct VestingFund {
     pub epoch: ChainEpoch,
     pub amount: TokenAmount,
 }
 
-/// Represents the vesting table state for the miner.
-/// It is a slice of (VestingEpoch, VestingAmount).
-/// The slice will always be sorted by the VestingEpoch.
-#[derive(Serialize_tuple, Deserialize_tuple, Default)]
-pub struct VestingFunds {
-    pub funds: Vec<VestingFund>,
+/// Represents the vesting table state for the miner. It's composed of a `head` (stored inline) and
+/// a tail (referenced by CID).
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+#[serde(transparent)]
+pub struct VestingFunds(Option<VestingFundsInner>);
+
+#[derive(Serialize_tuple, Deserialize_tuple, Debug, Clone)]
+struct VestingFundsInner {
+    // The "next" batch of vesting funds.
+    head: VestingFund,
+    // The rest of the vesting funds, if any.
+    tail: Cid, // Vec<VestingFund>
```

Comment:
This still could be `Optional<Cid>` so we could avoid storing the empty array CID when you get down to having nothing beyond the head left.

Reply:
It could be, but I didn't think that was worth optimizing. Miners should almost never be in that state.